File: solutions/public/p0091/p0091_s0.c
/* Solution to Euler Problem 91: Right Triangles with Integer Coordinates. */
#include "runner.h"

/* Euclidean gcd, used to reduce a vector to its primitive lattice direction; O(log min(a, b)). */
static int gcd(int a, int b) {
    while (b) {
        int t = b;
        b = a % b;
        a = t;
    }
    return a;
}

/*
 * Classify each right triangle by where its right angle sits. The origin contributes exactly
 * 3 * N^2 axis-aligned cases. For a right angle at non-origin P = (x, y) the perpendicular to OP
 * has primitive step (-y/m, x/m) with m = gcd(x, y); the count of lattice points reachable inside
 * the grid is bounded by both the horizontal room (x*m/y) and the vertical room (m*(N-y)/x), and
 * the minimum is the step count. Doubling covers the right angle at Q and the opposite direction.
 * Both divisions are exact because m divides x and y. Overall O(N^2 log N).
 */
const char *solve(int argc, char *argv[]) {
    static char _answer[32];
    int coordinate_limit = parse_int(argv[1]);

    long long triangles_at_p_or_q = 0;
    for (int x = 1; x <= coordinate_limit; x++) {
        for (int y = 1; y < coordinate_limit; y++) {
            int m = gcd(x, y);
            /* Cast to long long before multiplying so x*m cannot overflow 32 bits for large N. */
            long long a = (long long)x * m / y;
            long long b = (long long)m * (coordinate_limit - y) / x;
            long long mn = a < b ? a : b;
            triangles_at_p_or_q += mn;
        }
    }
    triangles_at_p_or_q *= 2;

    long long triangles_at_origin = 3LL * coordinate_limit * coordinate_limit;
    { snprintf(_answer, sizeof _answer, "%lld", (long long)(triangles_at_p_or_q + triangles_at_origin)); return _answer; }
}

File: solutions/public/p0091/p0091_s0.py
#!/usr/bin/env python3.14
# -*- coding: utf-8 -*-
""" Solution to Euler Problem 91: Right Triangles with Integer Coordinates [Level 6]. """
from __future__ import annotations

import math

from solver.runners import runner


@runner.main
def solve(*args: str) -> str:
    """Classify right triangles by the location of the right angle; O(N^2 log N).

    The origin contributes 3 * N^2 axis-aligned cases. For a right angle at non-origin P = (x, y),
    the primitive perpendicular step is (-y/m, x/m) with m = gcd(x, y); the reachable lattice points
    are bounded by horizontal room x*m/y and vertical room m*(N-y)/x, and the minimum counts them.
    Doubling covers the right angle at Q and the opposite direction; the gcd makes both divisions exact.
    """
    coordinate_limit = runner.parse_int(args[0])

    triangles_at_p_or_q = sum(
        (
            min(x * m // y, m * (coordinate_limit - y) // x)
            for x in range(1, coordinate_limit + 1)
            for y in range(1, coordinate_limit)
            for m in [math.gcd(x, y)]
        )
    )
    triangles_at_p_or_q *= 2
    triangles_at_origin = 3 * coordinate_limit**2
    return str(triangles_at_p_or_q + triangles_at_origin)


if __name__ == "__main__":
    raise SystemExit(solve())
